### ISSUE TITLE
dmatrix does not work with categorical variables

### ISSUE THREAD

**Issue by fccoelho** [pydata/patsy#68]
The following example:

```python
import pandas as pd
import patsy

data = pd.DataFrame({'a':[0,1,0,1,0,1,0,1,0,1]})
data.a = data.a.astype("category")
dmatrix("C(a)", data)
```

leads to this error:

```python
usr/local/lib/python3.4/dist-packages/patsy/highlevel.py in dmatrix(formula_like, data, eval_env, NA_action, return_type)
    276     eval_env = EvalEnvironment.capture(eval_env, reference=1)
    277     (lhs, rhs) = _do_highlevel_design(formula_like, data, eval_env,
--> 278                                       NA_action, return_type)
    279     if lhs.shape[1] != 0:
    280         raise PatsyError("encountered outcome variables for a model "

/usr/local/lib/python3.4/dist-packages/patsy/highlevel.py in _do_highlevel_design(formula_like, data, eval_env, NA_action, return_type)
    150         return iter([data])
    151     builders = _try_incr_builders(formula_like, data_iter_maker, eval_env,
--> 152                                   NA_action)
    153     if builders is not None:
    154         return build_design_matrices(builders, data,

/usr/local/lib/python3.4/dist-packages/patsy/highlevel.py in _try_incr_builders(formula_like, data_iter_maker, eval_env, NA_action)
     55                                        formula_like.rhs_termlist],
     56                                       data_iter_maker,
---> 57                                       NA_action)
     58     else:
     59         return None

/usr/local/lib/python3.4/dist-packages/patsy/build.py in design_matrix_builders(termlists, data_iter_maker, NA_action)
    658                                                    factor_states,
    659                                                    data_iter_maker,
--> 660                                                    NA_action)
    661     # Now we need the factor evaluators, which encapsulate the knowledge of
    662     # how to turn any given factor into a chunk of data:

/usr/local/lib/python3.4/dist-packages/patsy/build.py in _examine_factor_types(factors, factor_states, data_iter_maker, NA_action)
    427                     cat_sniffers[factor] = CategoricalSniffer(NA_action,
    428                                                               factor.origin)
--> 429                 done = cat_sniffers[factor].sniff(value)
    430                 if done:
    431                     examine_needed.remove(factor)

/usr/local/lib/python3.4/dist-packages/patsy/categorical.py in sniff(self, data)
    169         # fastpath to avoid doing an item-by-item iteration over boolean
    170         # arrays, as requested by #44
--> 171         if hasattr(data, "dtype") and np.issubdtype(data.dtype, np.bool_):
    172             self._level_set = set([True, False])
    173             return True

/usr/local/lib/python3.4/dist-packages/numpy/core/numerictypes.py in issubdtype(arg1, arg2)
    761     """
    762     if issubclass_(arg2, generic):
--> 763         return issubclass(dtype(arg1).type, arg2)
    764     mro = dtype(arg2).type.mro()
    765     if len(mro) > 1:

TypeError: data type not understood
```

This must be a bug. Is there any workaround?

**Comment by njsmith** [comment]
Yeah, pandas totally redid their categorical stuff in an incompatible way
since the last Patsy release.
.
Until the next release lands, workarounds include using the latest version
from master (this should work so if it doesn't please speak up), or
avoiding use of pandas's categorical pseudo-dtype when passong data into
patsy. (E.g. your example works fine if you comment out the astype line.)

**Comment by njsmith** [comment]
Fixed by 0.4.0
